src/style-spec/validate/validate_expression.js

```diff
@@ -41,3 +44,31 @@ export default function validateExpression(options: any): Array<ValidationError>
 
     return [];
 }
+
+export function disallowedFilterParameters(e: Expression, options: any): Array<ValidationError> {
+    const disallowedParameters = new Set([
+        'zoom',
+        'feature-state',
+        'pitch',
+        'distance-from-center'
+    ]);
+    for (const param of options.valueSpec.expression.parameters) {
+        disallowedParameters.delete(param);
+    }
+
+    if (disallowedParameters.size === 0) {
+        return [];
+    }
+    const errors = [];
+
+    if (e instanceof CompoundExpression) {
+        if (disallowedParameters.has(e.name)) {
+            return [new ValidationError(options.key, options.value, `["${e.name}"] expression is not supported in a filter for a ${options.object.type} layer with id: ${options.object.id}`)];
```

Comment:
Stylistic question here! Do we prefer to put the correct-use-case return in a nested value and error returns (or throws) at the end? I usually do it the other way around. In any case I'd like to be consistent!

This applies to line 25 too.

Reply:
I don't think we have a hard rule. I think doing it a way that makes it easier to read a particular function is better. I personally like error handling/validation at the top and doing the meat of the logic after.